File: 19_RemoveNthNodeFromEndOfList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        
        dummy = head
        count = 0
        
        while(dummy):
            count += 1
            dummy = dummy.next
            
        #print(count)
        
        countOfNodeToBeRemoved = count - n
        
        if countOfNodeToBeRemoved < 0:
            return None
        elif countOfNodeToBeRemoved == 0:
            head = head.next
            return head
        
        dummy = head
        prev = head
        while(countOfNodeToBeRemoved):
            prev = dummy
            dummy = dummy.next
            countOfNodeToBeRemoved -= 1
        
        prev.next = dummy.next
        
        return head
        
        

File: 19_RemoveNthNodeFromEndOfList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = head;
        int count = 0;
        
        while(temp != NULL)
        {
            temp = temp->next;
            count++;
        }
        
        //cout << count << endl;
        int nodeToBeRemoved = count - n;
        
        if(nodeToBeRemoved < 0)
        {
            head = NULL;
            return head;
        }
        else if(nodeToBeRemoved == 0)
        {
            head = head->next;
            return head;
        }
        
        temp = head;
        ListNode *dummy;
        
        while(nodeToBeRemoved)
        {
            dummy = temp;
            temp = temp->next;
            nodeToBeRemoved--;
            
        }
        
        if(dummy != NULL)
        {
            dummy->next = temp->next;
        }
        
        return head;
    }
};